sql/core/v1.2.1/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcFilters.scala

```diff
@@ -143,145 +115,231 @@ private[sql] object OrcFilters extends OrcFiltersBase {
       case And(left, right) =>
-        // At here, it is not safe to just convert one side and remove the other side
-        // if we do not understand what the parent filters are.
-        //
-        // Here is an example used to explain the reason.
-        // Let's say we have NOT(a = 2 AND b in ('1')) and we do not understand how to
-        // convert b in ('1'). If we only convert a = 2, we will end up with a filter
-        // NOT(a = 2), which will generate wrong results.
-        //
-        // Pushing one side of AND down is only safe to do at the top level or in the child
-        // AND before hitting NOT or OR conditions, and in this case, the unsupported predicate
-        // can be safely removed.
-        val leftBuilderOption =
-          createBuilder(dataTypeMap, left, newBuilder, canPartialPushDownConjuncts)
-        val rightBuilderOption =
-          createBuilder(dataTypeMap, right, newBuilder, canPartialPushDownConjuncts)
-        (leftBuilderOption, rightBuilderOption) match {
-          case (Some(_), Some(_)) =>
-            for {
-              lhs <- createBuilder(dataTypeMap, left,
-                builder.startAnd(), canPartialPushDownConjuncts)
-              rhs <- createBuilder(dataTypeMap, right, lhs, canPartialPushDownConjuncts)
-            } yield rhs.end()
-
-          case (Some(_), None) if canPartialPushDownConjuncts =>
-            createBuilder(dataTypeMap, left, builder, canPartialPushDownConjuncts)
-
-          case (None, Some(_)) if canPartialPushDownConjuncts =>
-            createBuilder(dataTypeMap, right, builder, canPartialPushDownConjuncts)
-
-          case _ => None
+        actionType match {
+          case t @ TrimUnconvertibleFilters(canPartialPushDownConjuncts) =>
+            // At here, it is not safe to just keep one side and remove the other side
+            // if we do not understand what the parent filters are.
+            //
+            // Here is an example used to explain the reason.
+            // Let's say we have NOT(a = 2 AND b in ('1')) and we do not understand how to
+            // convert b in ('1'). If we only convert a = 2, we will end up with a filter
+            // NOT(a = 2), which will generate wrong results.
+            //
+            // Pushing one side of AND down is only safe to do at the top level or in the child
+            // AND before hitting NOT or OR conditions, and in this case, the unsupported
+            // predicate can be safely removed.
+            val lhs = performAction(t, left)
+            val rhs = performAction(t, right)
+            (lhs, rhs) match {
+              case (Some(l), Some(r)) => Some(And(l, r))
+              case (Some(_), None) if canPartialPushDownConjuncts => lhs
+              case (None, Some(_)) if canPartialPushDownConjuncts => rhs
+              case _ => None
+            }
+          case BuildSearchArgument(builder) =>
+            builder.startAnd()
+            updateBuilder(left, builder)
```

Comment:
nit: to be consistent, we should call `performAction(action, left)`. It's clearer that `performAction` is recursive.